pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_sysprep.go

```diff
@@ -79,3 +84,74 @@ func BootstrapSysPrep(
 
 	return configSpec, customSpec, nil
 }
+
+func convertTo(from *vmopv1_prep.Sysprep, bootstrapData BootstrapData) *vimTypes.CustomizationSysprep {
+	sysprepCustomization := &vimTypes.CustomizationSysprep{}
+
+	if from.GUIUnattended != nil {
+		sysprepCustomization.GuiUnattended = vimTypes.CustomizationGuiUnattended{
+			Password: &vimTypes.CustomizationPassword{
+				Value:     *bootstrapData.SysprepPassword,
+				PlainText: true,
+			},
+			TimeZone:       from.GUIUnattended.TimeZone,
+			AutoLogon:      from.GUIUnattended.AutoLogon,
+			AutoLogonCount: from.GUIUnattended.AutoLogonCount,
+		}
+	}
+
+	if from.UserData != nil {
+		sysprepCustomization.UserData = vimTypes.CustomizationUserData{
+			FullName: from.UserData.FullName,
+			OrgName:  from.UserData.OrgName,
+		}
+		// In the case of a VMI with volume license key, this might not be set.
+		// Hence add a check to see if the productID is set to empty.
+		if bootstrapData.SysprepProductID != nil {
+			sysprepCustomization.UserData.ProductId = *bootstrapData.SysprepProductID
+		}
+	}
+
+	sysprepCustomization.GuiRunOnce = &vimTypes.CustomizationGuiRunOnce{
+		CommandList: from.GUIRunOnce.Commands,
+	}
+
+	if from.Identification != nil {
+		sysprepCustomization.Identification = vimTypes.CustomizationIdentification{
+			JoinWorkgroup: from.Identification.JoinWorkgroup,
+			JoinDomain:    from.Identification.JoinDomain,
+			DomainAdmin:   from.Identification.DomainAdmin,
+			DomainAdminPassword: &vimTypes.CustomizationPassword{
+				Value:     *bootstrapData.SysprepDomainPassword,
```

Comment:
Will this be `nil` when `Identification.JoinDomain` is `""`?

Reply:
Removed the pointer to avoid an accidental NPE, also moved the logic to add it only if the value is set in `bootstrapData`.